lib/fs.js

```diff
@@ -827,117 +808,192 @@ fs.ftruncate = function(fd, len, callback) {
   if (typeof len === 'function') {
     callback = len;
     len = 0;
-  } else if (len === undefined) {
+  } else if (len == null) {
     len = 0;
   }
-  var req = new FSReqWrap();
+  if (!Number.isInteger(fd) || fd < 0) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'fd',
+                               'unsigned integer');
+  }
+
+  if (!Number.isInteger(len) || len < 0) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'len',
+                               'unsigned integer');
+  }
+  const req = new FSReqWrap();
   req.oncomplete = makeCallback(callback);
   binding.ftruncate(fd, len, req);
 };
 
 fs.ftruncateSync = function(fd, len) {
-  if (len === undefined) {
+  if (len == null)
     len = 0;
+  if (!Number.isInteger(fd) || fd < 0) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'fd',
+                               'unsigned integer');
+  }
+
+  if (!Number.isInteger(len) || len < 0) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'len',
+                               'unsigned integer');
```

Comment:
Can't a lot of these checks be done in c++? it would be a nice performance win.

Reply:
Pulled many of these *out* of C++ in order to get better consistency with the move to internal/errors and consistency with where the type checking is happening. As it is, the type checking is rather... inconsistent. Will look at it again, tho.